File: Problem120/Solution.go
package main

import (
	"math"
)

func minimumTotal(triangle [][]int) int {
	n := len(triangle)
	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, n)
	}

	for i := 0; i < n; i++ {
		dp[0][i] = triangle[n-1][i]
	}

	k := 0
	for i := n - 2; i >= 0; i-- {

		for j := 0; j < i+1; j++ {
			if dp[k][j] > dp[k][j+1] {
				dp[k+1][j] = dp[k][j+1] + triangle[i][j]
			} else {
				dp[k+1][j] = dp[k][j] + triangle[i][j]
			}
		}
		k++
	}
	return dp[n-1][0]
}

File: Problem120/Solution.java
/**
 * @author tgorthi
 * @since Jun 2020
 */

import java.util.ArrayList;
import java.util.List;

public class Solution
{
    public int minimumTotal(List<List<Integer>> triangle)
    {
        int n = triangle.size();
        List<List<Integer>> dp = new ArrayList<>();
        dp.add(triangle.get(n - 1));
        int k = 0;
        for (int i = n - 2; i >= 0; i--)
        {
            List<Integer> list = triangle.get(i);
            List<Integer> localMax = new ArrayList<>();
            for (int j = 0; j < list.size(); j++)
            {
                int min = Math.min(dp.get(k).get(j), dp.get(k).get(j + 1));
                localMax.add(min + list.get(j));
            }
            dp.add(localMax);
            k++;
        }
        return dp.get(n - 1).get(0);
    }
}